File: app/src/main/java/com/guerinet/mymartlet/ui/dialog/list/TermDialogHelper.java
/*
 * Copyright 2014-2017 Julien Guerinet
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.guerinet.mymartlet.ui.dialog.list;

import android.content.Context;

import com.guerinet.mymartlet.App;
import com.guerinet.mymartlet.model.Semester;
import com.guerinet.mymartlet.model.Term;
import com.guerinet.mymartlet.util.dagger.prefs.DefaultTermPref;
import com.guerinet.mymartlet.util.dagger.prefs.RegisterTermsPref;
import com.guerinet.suitcase.analytics.GAManager;
import com.guerinet.suitcase.dialog.SingleListInterface;
import com.raizlabs.android.dbflow.sql.language.SQLite;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

import javax.inject.Inject;

/**
 * {@link SingleListInterface} implementation for a list of terms
 * @author Julien Guerinet
 * @since 2.0.0
 */
public abstract class TermDialogHelper implements SingleListInterface {
    /**
     * App context
     */
    @Inject
    protected Context context;

    @Inject
    protected GAManager ga;

    @Inject
    DefaultTermPref defaultTermPref;

    @Inject
    RegisterTermsPref registerTermsPref;
    /**
     * List of {@link Term}s to choose from
     */
    private final List<Term> terms;
    /**
     * Currently selected {@link Term}
     */
    private final Term currentTerm;

    /**
     * Default Constructor
     *
     * @param context App context
     */
    public TermDialogHelper(Context context, Term currentTerm, boolean registration) {
        App.Companion.component(context).inject(this);

        ga.sendScreen("Change Semester");

        //Use the default currentTerm if no currentTerm was sent
        this.currentTerm = currentTerm == null ? defaultTermPref.getTerm() : currentTerm;

        terms = new ArrayList<>();
        if (!registration) {
            List<Semester> semesters = SQLite.select()
                    .from(Semester.class)
                    .queryList();

            // We are using the user's existing terms
            for (Semester semester : semesters) {
                terms.add(semester.getTerm());
            }
        } else {
            // We are using the registration terms
            terms.addAll(registerTermsPref.getTerms());
        }

        //Sort them chronologically
        Collections.sort(terms, new Comparator<Term>() {
            @Override
            public int compare(Term term, Term term2){
                return term.isAfter(term2) ? -1 : 1;
            }
        });
    }

    @Override
    public int getCurrentChoice() {
        return terms.indexOf(currentTerm);
    }

    @Override
    public String[] getChoices() {
        String[] titles = new String[terms.size()];
        for (int i = 0; i < terms.size(); i ++) {
            titles[i] = terms.get(i).getString(context);
        }

        return titles;
    }

    @Override
    public void onChoiceSelected(int position) {
        onTermSelected(terms.get(position));
    }

    /**
     * Called when a {@link Term} has been selected
     *
     * @param term The selected {@link Term}
     */
    public abstract void onTermSelected(Term term);
}

File: app/src/main/java/com/guerinet/mymartlet/ui/dialog/list/TermDialogHelper.kt
/*
 * Copyright 2014-2018 Julien Guerinet
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.guerinet.mymartlet.ui.dialog.list

import android.content.Context
import android.util.Pair
import com.guerinet.mymartlet.R
import com.guerinet.mymartlet.model.Semester
import com.guerinet.mymartlet.model.Term
import com.guerinet.mymartlet.util.prefs.DefaultTermPref
import com.guerinet.mymartlet.util.prefs.RegisterTermsPref
import com.guerinet.suitcase.analytics.GAManager
import com.guerinet.suitcase.dialog.singleListDialog
import com.raizlabs.android.dbflow.kotlinextensions.from
import com.raizlabs.android.dbflow.sql.language.SQLite
import org.koin.standalone.KoinComponent
import org.koin.standalone.inject

/**
 * Displays a list of terms
 * @author Julien Guerinet
 * @since 2.0.0
 */
class TermDialogHelper(context: Context, currentTerm: Term?, registration: Boolean,
        onTermSelected: ((Term) -> Unit)) : KoinComponent {

    private val ga by inject<GAManager>()

    private val defaultTermPref by inject<DefaultTermPref>()

    private val registerTermsPref by inject<RegisterTermsPref>()

    private val terms by lazy {
        if (!registration) {
            // We are using the user's existing terms
            SQLite.select()
                    .from(Semester::class)
                    .queryList()
                    .map { it.term }
        } else {
            // We are using the registration terms
            registerTermsPref.terms.toList()
        }
                .sortedWith(kotlin.Comparator { o1, o2 -> if (o1.isAfter(o2)) -1 else 1 })
                .map { Pair(it, it.getString(context)) }
    }

    init {
        ga.sendScreen("Change Semester")

        terms.map { it.second }.toTypedArray()

        val term = currentTerm ?: defaultTermPref.getTerm()

        // Use the default currentTerm if no currentTerm was sent
        val currentChoice = terms.indexOfFirst { it.first == term }

        val choices = terms.map { it.second }.toTypedArray()

        context.singleListDialog(choices, R.string.title_change_semester, currentChoice) {
            onTermSelected(terms[it].first)
        }
    }
}
